File: lecture_notes/10 refactoring/parse-3.py
import json


def parse_shift(points):
    return [json.loads(point) if type(point) == str else point for point in points]


def shift(points, offset):
    for point in points:
        point['x'] += offset['x']
        point['y'] += offset['y']
    
    return points


polyline = [
  { 'x': 0, 'y': 0 },
  { 'x': 10, 'y': 10 },
  '{ "x": 20, "y": 20 }',
  { 'x': 30, 'y': 30 }
]

parser = parse_shift(polyline)
print(shift(parser, {'x': 10, 'y': -5}))

File: lecture_notes/10 refactoring/parse-3.js
"use strict";

const shift =
  ({ a, b }) =>
  ({ x, y }) => ({ x: x + a, y: y + b });

const checkType = (line) =>
  typeof line !== "object" ? JSON.parse(line) : line;

const offset = shift({ a: 10, b: -5 });

const polyline = [
  { x: 0, y: 0 },
  { x: 10, y: 10 },
  '{ "x": 20, "y": 20 }',
  { x: 30, y: 30 },
];

const checkedType = polyline.map(checkType);
const path = checkedType.map(offset);
console.log({ path });
